src/settings.lisp

```diff
@@ -18,8 +18,54 @@
 
 (in-package #:coalton-impl/settings)
 
+;;; Configuration reaches the Coalton compiler in 3 ways.
+;;;
+;;;    1. When Coalton is compiled, several environment variables are
+;;;    inquired. These are
+;;;
+;;;        - COALTON_ENV: Whether to be in development or release mode.
+;;;
+;;;        - COALTON_DISABLE_SPECIALIZATION: Whether to disable
+;;;          function specialization.
+;;;
+;;;        - COALTON_HEURISTIC_INLINING: Whether to enable automatic
+;;;          inlining.
+;;;
+;;;    2. When Coalton is compiled, configuration on properties of the
+;;;    :COALTON-CONFIG symbol are inquired. These are secondary to the
+;;;    environment variables.
+;;;
+;;;    3. When the Coalton compiler is run, several special variables
+;;;    control how the code is compiled. The values of the special
+;;;    variables are affected by the aforementioned configuration.
+
+(eval-when (:compile-toplevel :load-toplevel :execute)
+  (defvar *config-keys*
+    '(:compiler-mode                    ; [string] compiler mode
+      :print-unicode                    ; [boolean] print unicode?
+      :perform-specialization           ; [boolean] use specializations?
+      :perform-inlining                 ; [boolean] automatic inlining?
+      :emit-type-annotations            ; [boolean] emit type annotations?
+      :print-types                      ; [boolean] print types when compiling?
+      )
+    "Valid configuration keys that can be (SETF GET) on the user configuration variable :COALTON-CONFIG.")
+
+  (defun config (key &key (default nil defaultp))
+    "Get the user configuration associated with the key KEY. If it's not
+found, return DEFAULT instead."
+    (assert defaultp () "A default must be supplied to CONFIG.")
+    (cond
+      ((member key *config-keys*)
+       (get ':coalton-config key default))
+      (t
+       (warn "Unknown Coalton configuration key: ~A" key)
+       default))))
+
+
+;;; Compiler and runtime configuration
+
 (declaim (type boolean *coalton-print-unicode*))
-(defvar *coalton-print-unicode* t
+(defvar *coalton-print-unicode* (config ':print-unicode :default t)
   "Whether to print coalton info using unicode symbols")
 
 (defun coalton-release-p ()
```

Comment:
This should maybe mention https://github.com/coalton-lang/coalton/blob/main/docs/coalton-lisp-interop.md
which I believe is the only thorough documentation on the difference between the two modes.